### PR TITLE
Fix the hash prop on Links

### LINKED ISSUE #2216: Fix <Link hash>
I have two links, one created with `<Link to='/faq' hash='#whatis'>` and the other one created with `<a href='/faq#whereq'>` that look as following when inspecting element on rendered document:

```
<a class="" href="/faq#whatis" data-reactid=".0.2.0.0.0.1.0">...</a>
<a href="/faq#whereq" data-reactid=".0.2.0.0.0.2.0">...</a>
```

When I click on the second one, hash is populated in the location object and I'm being taken to the fragment. When I click on the first one, I'm being taken to `\faq`, hash is not populated in the location object and I'm not being taken to the fragment. Any idea why is that?

### PR DESCRIPTION
**Pull request by timdorr**
Fixes #2216

### CODE REVIEW

File: modules/__tests__/Link-test.js [hunk 1]
```diff
@@ -289,23 +289,27 @@ describe('A <Link>', function () {
           // just here to make sure click handlers don't prevent it from happening
         }
         render() {
-          return <Link to="/hello" onClick={(e) => this.handleClick(e)}>Link</Link>
+          return <Link to="/hello" hash="#world" onClick={(e) => this.handleClick(e)}>Link</Link>
```
Comment (mjackson) [inline modules/__tests__/Link-test.js]: Let's add a `query` prop here as well, to be a bit more stringent.

**Comment by timdorr** [comment]
Oh shit, merged too quickly. I was waiting on tests to pass. I'll follow up on the comments.

**Comment by aszwemin** [comment]
I think that will still not jump to the fragment, made similar fix in my local branch initially, but it only solved url.

**Comment by mjackson** [comment]
np :) In general, it's probably best if we all wait to get at least one :+1: from another collab before merging.

**Comment by mjackson** [comment]
@aszwemin Jumping to the fragment is a scrolling problem, which is a separate issue.

**Comment by timdorr** [comment]
It's because the URL is being constructed the wrong way. The hash fragment and query string are out of order. I'm going to submit an issue on `history` about it.

File: modules/__tests__/Link-test.js (hunk 1 above)
Comment (timdorr) [inline modules/__tests__/Link-test.js]: Fixed in 81cccce

Comment (timdorr) [inline modules/__tests__/Link-test.js]: And I'll wait for the +1 next time :smile:

**Comment by taion** [comment]
Do we need any sanity checks here when using hash history?

**Comment by timdorr** [comment]
Well, we're now testing another aspect of it (I'd love to get some coverage reports generated...) but I think it's warranted.

But once they bug in `history` is fixed, this line: 
```js
expect(spy).toHaveBeenCalledWith({ you: 'doing?' }, '/hello#world?how=are')
```
changes to:
```js
expect(spy).toHaveBeenCalledWith({ you: 'doing?' }, '/hello?how=are#world')
```
So, it's not so much of a sanity check. It's actually testing the valid construction of a URL.

**Comment by aszwemin** [comment]
@mjackson I hacked something like this locally, not sure if it's a good enough solution though:

```
 87         if (this.props.hash) {                                                  
 88             container = _react2['default'].findDOMNode(this);                   
 89             element = document.getElementById(this.props.hash.slice(1));  
 90             if (element) window.scrollTo(container.offsetLeft, element.offsetTop);           
 91         } 
```

Happy to make a pr if this is what you had in mind, but not sure how generic it would be

**Pull request merged**